src/sync/mutex.rs

```diff
@@ -94,15 +94,79 @@ impl<T: ?Sized> Mutex<T> {
                 mutex: self,
             })),
             Err(TryLockError::WouldBlock) => panic!("mutex state out of sync"),
-        }
+        };
+
+        // We need to let other threads in here so they may fail a `try_lock`
+        thread::switch();
+
+        result
     }
 
     /// Attempts to acquire this lock.
     ///
     /// If the lock could not be acquired at this time, then Err is returned. This function does not
     /// block.
     pub fn try_lock(&self) -> TryLockResult<MutexGuard<T>> {
-        unimplemented!()
+        let me = ExecutionState::me();
+
+        let mut state = self.state.borrow_mut();
+        trace!(holder=?state.holder, waiters=?state.waiters, "trying to acquire mutex {:p}", self);
+
+        if let Some(holder) = state.holder {
+            if holder == me {
+                panic!("deadlock! task {:?} tried to acquire a Mutex it already holds", me);
+            }
+        }
+
+        // We don't need a context switch here. There are two cases to analyze.
+        // * Consider that we manage to acquire the lock, but that there is another thread that
+        //   also wants to acquire. At the last context switch (where we were chosen), the other
+        //   thread must have been already runnable and could have been chosen by the scheduler.
+        //   Then that other thread's acquire has a context switch that allows us to run into the
+        //   `WouldBlock` case.
+        // * Consider that we run into the `WouldBlock` case, but that there is another thread
+        //   that wants to release. At the last context switch (where we were chosen), the other
+        //   thread must have been already runnable and could have been chosen by the scheduler.
+        //   Then that other thread's release has a context switch that allows us to acquire the
+        //   lock.
+
+        let result = if state.holder.is_none() {
+            state.holder = Some(me);
+
+            trace!("acquired mutex {:p}", self);
+
+            // Re-block all other waiting threads, since we won the race to take this lock
+            for tid in state.waiters.iter() {
+                ExecutionState::with(|s| s.get_mut(tid).block());
+            }
+            // Update acquiring thread's clock with the clock stored in the Mutex
```

Comment:
You need to do the vector clock update even in the failing case, because there's a causal dependency: if the `try_lock` fails, the current thread T1 *knows* that the thread T2 that owns the lock is in its critical section, and therefore T1 has a causal dependency on everything that happened before in T2 (which is recorded in the lock's clock, at line 83).

Reply:
Actually, there's even more to it.  In the failing try_lock case, you also have to update the vector clock stored in the Mutex, because future threads that manage to acquire the lock have a causal dependency on this failed try_lock.